pyquil/quilatom.py

```diff
@@ -537,6 +537,51 @@ def __neg__(self) -> "Mul":
     def _substitute(self, d: Any) -> ExpressionDesignator:
         return self
 
+    def _evaluate(self) -> np.complex128:
+        """
+        Attempts to evaluate the expression to by simplifying it to a complex number.
+
+        Expression simplification can be slow, especially for large recursive expressions.
+        This method will raise a ValueError if the expression cannot be simplified to a complex
+        number.
+        """
+        expr = quil_rs_expr.Expression.parse(str(self))
+        expr.simplify()  # type: ignore
+        if not expr.is_number():
+            raise ValueError(f"Cannot evaluate expression {self} to a number. Got {expr}.")
+        return np.complex128(expr.to_number())
+
+    def __float__(self) -> float:
+        """
+        Returns a copy of the expression as a float by attempting to simplify the expression.
+
+        Expression simplification can be slow, especially for large recursive expressions.
+        This cast will raise a ValueError if simplification doesn't result in a real number.
+        """
+        value = self._evaluate()
+        if value.imag != 0:
```

Comment:
Can we use `np.isclose` with a fairly moderate threshold here? Perhaps 1e-8?

Reply:
I'd thought about this too but decided it wasn't necessary; if the inputs are all real, `imag == 0.0` in all of the inputs and there are no complex operations I can think of that would accumulate a rounding error. So I'd be inclined to leave it unless you have a specific case that would trip it.

Reply:
@bramathon Any thoughts on the above? Just want to make sure before I merge as-is.